More unit tests

kuba-04 opened the pull request:
#4470

tcharding commented on bitcoin/src/crypto/key.rs:
```diff
@@ -1520,6 +1532,25 @@ mod tests {
+        let pk = "032e58afe51f9ed8ad3cc7897f634d881fdbe49a81564629ded8156bebd2ffd1af"
+            .parse::<PublicKey>()
+            .unwrap();
+        let result = pk.p2wpkh_script_code();
+        assert!(result.is_ok());
+
+        let script = result.unwrap();
```
```rust
let script = pk.p2wpkh_script_code().unwrap();
```

tcharding commented on bitcoin/src/crypto/key.rs:
```diff
@@ -1841,3 +1872,33 @@ mod tests {
 }
+
+#[cfg(all(test, feature = "std"))]
+mod std_tests {
```
This whole module is very unusual, can you explain what you are trying to achieve?

tcharding reviewed (commented):
Thanks for the contribution. This will need some work still to get in. You've got 4 patches all claiming to resolve the same issue. Can you just put a single line in the PR description that does that and not in the individual patches?

Also, perhaps explain what you are trying to achieve in each patch then I can do a more meaningful review for you.

Thanks!

kuba-04 commented on bitcoin/src/crypto/key.rs:
```diff
@@ -1468,6 +1473,13 @@ mod tests {
+            secp256k1::SecretKey::from_str("c28a9f80738efe59906e5a1b4c4021e8c2d8e0b1c9c3d1e8b8a7b5c4f2e1d8c7").unwrap(),
+            NetworkKind::Main);
+        assert!(!sk.compressed);
+        assert_eq!(&sk.to_wif(), "5KHxtARu5yRpYLc3WZRApb8DbPGpcDQ2L3uufJ99izMamS2dXLQ");
```
You are right, this is not clear. We want to assert on the same value so the test is at least readable

kuba-04 commented on bitcoin/src/crypto/key.rs:
```diff
@@ -1520,6 +1532,25 @@ mod tests {
+        assert_eq!(script_bytes[2], 0x14);
+        assert_eq!(script_bytes[23], OP_EQUALVERIFY.to_u8());
+        assert_eq!(script_bytes[24], OP_CHECKSIG.to_u8());
+    }
```
I wanted to cover this method:
```rust
pub fn p2wpkh_script_code(&self) -> Result<ScriptBuf, UncompressedPublicKeyError> {
        let key = CompressedPublicKey::try_from(*self)?;
        Ok(key.p2wpkh_script_code())
    }
```
Again, this was for improving the coverage, but doesn't have any value. It's rather checking underneath script.p2wpkh_script_code:
```rust
pub fn p2wpkh_script_code(wpkh: WPubkeyHash) -> ScriptBuf {
    Builder::new()
        .push_opcode(OP_DUP)
        .push_opcode(OP_HASH160)
        .push_slice(wpkh)
        .push_opcode(OP_EQUALVERIFY)
        .push_opcode(OP_CHECKSIG)
        .into_script()
}
```
removing. I think those methods that are not covered by any test, are not requiring testing, basically not every method requires a test.

kuba-04 commented:
> Thanks for the contribution. This will need some work still to get in. You've got 4 patches all claiming to resolve the same issue. Can you just put a single line in the PR description that does that and not in the individual patches?
> 
> Also, perhaps explain what you are trying to achieve in each patch then I can do a more meaningful review for you.
> 
> Thanks!

I created a draft so it will be easier for me to look at the changes and I was planning to go through it this morning and do some cleanup before I submit. But you were faster, which was great for me because your comments are very valuable, so I could reflect on the changes, removed/fixed some and squashed the commits. Most likely this one commit should not close the issue at all, it just adds some not very relevant tests so I changed `Closes` into `References`.  

Actually I will take a chance to ask you for advice. I just finished reading Mastering Bitcoin and my plan with this issue was to slowly add more tests to put in practice what I learned. I think that book doesn't cover enough detail to be valuable contributor to prod code, and I need to put more work/reading in something, perhaps Programming Bitcoin. For now, I wanted to start with improving the test coverage. Probably went to the wrong place as those I added are not valuable, majority of important code is already covered there. Given your knowledge about the project, do you have any suggestions on which areas I could focus on with some unit tests?

tcharding commented:
Great! We definitely could use more testers. So I can help you better perhaps give me a quick response to these questions if you feel like it:

- (I assume I know the answer to 'why Bitcoin')
- Why Rust?
- Why `rust-bitcoin`?
- What are you hoping to learn? (I'm more focused on software development than Bitcoin as such, there are better teachers for Bitcoin stuff than my reviews).

Just short answers is good so I get a clue. Thanks

kuba-04 commented:
thanks for not discouraging (yet!):
1. At some point in life you might want to remove as many useless and harmful things as possible and replace them with something you believe is good (call it decluttering);

2. It became a pleasure to pair-program with the compiler. But more seriously, mainly due to the projects that are being built in Rust, the interesting use cases (I come from Java background btw). Also, I see the positive dynamic in the ecosystem, improvements, new features, you have to keep learning and (hopefully) will never get bored;

3. I though that spending time here will give me the best understanding as it is the closest to the core. I see this project as a maintenance or plumbing, something which is not exciting, but difficult and necessary, someone has to do this - in this sense it is even more rewarding than the 'flashy' projects that come and go. But I might be wrong with this choice, maybe I should start with other projects that are closer to the "end product" (wallets) before getting into the details that are here? 

4. I hope to get better at Rust and Bitcoin at the same time (killing two birds with one stone basically).

tcharding commented:
Awesome man, thanks. Welcome.

tcharding commented:
If you want to do `serde` stuff do it in a different PR. If you want to push forward with this PR by all means do so. I was just trying to help narrow the scope for you since this one is a bit wide open at the moment and its not obvious during review what you are trying to achieve.

tcharding reviewed (approved):
ACK a04b71795cedca97e22a9bad622f057393d39707

kuba-04 commented:
ok, I will focus on `serde` from now on

tcharding commented:
Face palm, I didn't link you to the task, here it is: https://github.com/rust-bitcoin/rust-bitcoin/discussions/4932

(FYI we have started using discussions instead of issues for more involved tasks.) It will take multiple PRs to get this done so you could pick some way to limit scope and then go for it, for example you could look at `units/tests/serde.rs`, work out what is covered and what is not then do a few PRs something like (these are just ideas):

- fuzz test serde logic for types in `units`.
- roundtrip _both_ human-readable and non-human-readable for all types in `units`
- regression test ...

I'm not sure _exactly_ what needs doing, but starting with `units` we can pin it down and then do the same later in `primitives`, and eventually `bitcoin`.

tcharding commented:
Is this live still? Will close in two weeks unless requested otherwise.

(We are planning on [migrating away from GitHub](https://github.com/rust-bitcoin/rust-bitcoin/discussions/5714) so it is an opportune moment to clean up some stale PRs/issues.)
